csharp/ql/src/experimental/Security Features/JsonWebTokenHandler/JsonWebTokenHandlerLib.qll

```diff
@@ -0,0 +1,265 @@
+  CallableAlwaysReturnsTrue() {
+    callableHasAReturnStmtAndAlwaysReturnsTrue(this)
+    or
+    lambdaExprReturnsOnlyLiteralTrue(this)
+    or
+    exists(LambdaExpr le, Call call, CallableAlwaysReturnsTrue cat | this = le |
+      call = le.getExpressionBody() and
+      cat.getACall() = call
+    )
+  }
+}
+
+/**
+ * Holds if any exception being thrown by the callable is of type `System.ArgumentNullException`
+ * It will also hold if no exceptions are thrown by the callable
+ */
+predicate callableOnlyThrowsArgumentNullException(Callable c) {
+  forall(ThrowElement thre | c = thre.getEnclosingCallable() |
+    thre.getThrownExceptionType().hasQualifiedName("System.ArgumentNullException")
+  )
+}
+
+/**
+ * A specialization of `CallableAlwaysReturnsTrue` that takes into consideration exceptions being thrown for higher precision.
+ */
+class CallableAlwaysReturnsTrueHigherPrecision extends CallableAlwaysReturnsTrue {
+  CallableAlwaysReturnsTrueHigherPrecision() {
+    callableOnlyThrowsArgumentNullException(this) and
+    (
+      forall(Call call, Callable callable | call.getEnclosingCallable() = this |
+        callable.getACall() = call and
+        callable instanceof CallableAlwaysReturnsTrueHigherPrecision
+      )
+      or
+      exists(LambdaExpr le, Call call, CallableAlwaysReturnsTrueHigherPrecision cat | this = le |
+        call = le.getExpressionBody() and
+        cat.getACall() = call
+      )
+    )
+  }
+}
+
+/**
+ * A property Write for the `IssuerValidator` property for `Microsoft.IdentityModel.Tokens.TokenValidationParameters`
+ */
+class TokenValidationParametersPropertyWriteToValidationDelegatedIssuerValidator extends PropertyWrite {
+  TokenValidationParametersPropertyWriteToValidationDelegatedIssuerValidator() {
+    exists(Property p, Class c |
+      c.hasQualifiedName("Microsoft.IdentityModel.Tokens.TokenValidationParameters")
+    |
+      p.getAnAccess() = this and
+      c.getAProperty() = p and
+      p.hasName("IssuerValidator")
+    )
+  }
+}
+
+/**
+ * A callable that returns a `string` and has a `string` as 1st argument
+ */
+private class CallableReturnsStringAndArg0IsString extends Callable {
+  CallableReturnsStringAndArg0IsString() {
+    this.getReturnType() instanceof StringType and
+    this.getParameter(0).getType().toString() = "String"
+  }
+}
+
+/**
+ * A Callable that always return the 1st argument, both of `string` type
+ */
+class CallableAlwaysReturnsParameter0 extends CallableReturnsStringAndArg0IsString {
+  CallableAlwaysReturnsParameter0() {
+    forall(ReturnStmt rs | rs.getEnclosingCallable() = this |
+      rs.getChild(0) = this.getParameter(0).getAnAccess()
+    ) and
+    exists(ReturnStmt rs | rs.getEnclosingCallable() = this)
+    or
+    exists(LambdaExpr le, Call call, CallableAlwaysReturnsParameter0 cat | this = le |
+      call = le.getExpressionBody() and
+      cat.getACall() = call
+    )
+    or
+    this.getBody() = this.getParameter(0).getAnAccess()
+  }
```

Comment:
Some of this logic is similar to what we have in `Callable::canReturn`. I think it would be better to use that one plus the special handling of lambdas.

Reply:
I tried the following code to use `Callable::canReturn`, but it was missing a couple of cases (test case lines 117 & 118):

```
  predicate callableHasAReturnStmtAndAlwaysReturnsTrue(Callable c) {
    c.getReturnType() instanceof BoolType and
    not callableMayThrowException(c) and
    exists( Expr e | c.canReturn(e) | e.getValue()="true") and
    not exists( Expr e | c.canReturn(e) | e.getValue()!="true") 
  }
```

Any ideas how could I fix this one?

Reply:
Awaiting for some feedback on how to improve it using `Callable::canReturn`. Thanks

Reply:
I think `@hvitved` suggested an improvement in https://github.com/github/codeql/pull/9693/files/5a7b6532a9b4e9150baaa1780f355ceda554eba4#r942153852

Reply:
I didn't notice it was the same code. Fixed